### PR TITLE
[GraphBolt] Subgraph sampler base datapipe

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
This is a sub-PR of #5901 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/graphbolt/subgraph_sampler.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    .. code:: python
+
+       for data in dp:
+           yield sampler_func(graph, data)
+
+    Parameters
+    ----------
+    dp : DataPipe
+        The datapipe.
+    graph : GraphStorage
```
Comment (frozenbugs) [inline python/dgl/graphbolt/subgraph_sampler.py:20]: the sampler_func should already have the information of the graph, we don't need to pass the graph to the graph as a parameter to it.

File: tests/python/pytorch/graphbolt/test_subgraph_sampler.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import dgl.graphbolt
+import torch
+import torchdata.datapipes as dp
+
+
+def test_SubgraphSampler():
+    graph = dgl.add_reverse_edges(dgl.rand_graph(20, 60))
+    sampler = dgl.dataloading.NeighborSampler([2, 2])
+
+    def sampler_func(graph, data):
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_subgraph_sampler.py:11]:
```suggestion
    def sampler_func(data):
```

File: tests/python/pytorch/graphbolt/test_subgraph_sampler.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import dgl
+import dgl.graphbolt
+import torch
+import torchdata.datapipes as dp
+
+
+def test_SubgraphSampler():
+    graph = dgl.add_reverse_edges(dgl.rand_graph(20, 60))
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_subgraph_sampler.py:8]: Can you try with https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/graph_storage/csc_sampling_graph.py?

Comment (BarclayII) [reply tests/python/pytorch/graphbolt/test_subgraph_sampler.py:8]: How should I generate a random undirected graph with `CSCSamplingGraph`?

Comment (BarclayII) [reply tests/python/pytorch/graphbolt/test_subgraph_sampler.py]: I'm using scipy now.

File: tests/python/pytorch/graphbolt/test_subgraph_sampler.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        adjs = []
+        seeds = data
+
+        for hop in range(2):
+            sg = graph.sample_neighbors(seeds, torch.LongTensor([2]))
+            seeds = sg.indices
+            adjs.insert(0, sg)
+        return seeds, data, adjs
+
+    minibatch_dp = dp.map.SequenceWrapper(torch.arange(10)).batch(2)
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/test_subgraph_sampler.py:21]: why not use existing `graphbolt.MinibatchSampler`?

Comment (BarclayII) [reply tests/python/pytorch/graphbolt/test_subgraph_sampler.py:41]: Added

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/subgraph_sampler.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    Parameters
+    ----------
+    dp : DataPipe
+        The datapipe.
+    sampler_func : callable
+        The subgraph sampling function.
+    """
+
+    def __init__(self, dp, sampler_func):
```
Comment (frozenbugs) [inline python/dgl/graphbolt/subgraph_sampler.py:24]:
```suggestion
    def __init__(self, data_pipe, sampler_func):
```

Comment (BarclayII) [reply python/dgl/graphbolt/subgraph_sampler.py]: No longer need any code for initialization.

**Pull request merged**